### PR TITLE
fence_eaton_ssh: new fence agent for Eaton ePDU G3 over SSH

### PR DESCRIPTION
**Pull request by SchoolGuy**
Since the Eaton SNMP fence is not feasible for the infrastructure I am responsible for, I have decided to implement this fence that will use the Eaton SSH/Serial interface.

I also took the liberty to reformat the README to 120 characters so it is easier readable when looked at in raw and to improve the Python typing experience so adding new fences is easier.

### CODE REVIEW

File: agents/eaton_ssh/README.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# fence_eaton_ssh
+
+This was tested with:
+
```
Comment (oalbrigt) [inline agents/eaton_ssh/README.md:4]: We usually have this info as a comment at the top of the agent (see fence_apc and other agents). I guess you could also mention it in the longdesc of the agent.

Comment (SchoolGuy) [reply agents/eaton_ssh/README.md:4]: Thanks. I will correct this with the next push.

**Comment by SchoolGuy** [comment]
Missing points:

- `set_power_status` must be tested manually.
- `reboot_cycle` must be implemented and tested.
- `sync_set_power` must be implemented and tested.

All other functionality is tested as described in the docstrings. An initial feeling from someone about the code would be helpful as I am getting close to completing this PR.

**Comment by SchoolGuy** [comment]
This PR is now ready for review. The new fence is tested as describe in the Python docstring.

File: lib/fencing.py.py [hunk 2]
```diff
@@ -33,6 +34,13 @@
 
 LOG_FORMAT = "%(asctime)-15s %(levelname)s: %(message)s"
 
+OPTIONS_TYPE = Dict[str, Any]
+SET_POWER_FN_TYPE = Callable[[Optional["fspawn"], OPTIONS_TYPE], Optional[NoReturn]]
```
Comment (oalbrigt) [inline lib/fencing.py.py:38]: Typing is new in 3.5, so we should not use it in the fencing library, as it needs to be backwards compatible Python 2.7. Ideally we should keep new agents compatible with most 3.x versions as well.

https://docs.python.org/3/library/typing.html

Comment (SchoolGuy) [reply lib/fencing.py.py:38]: If I would have known that Python 2.7 compatibility is required, I would have not done these changes of course. When do you plan on dropping Python 2.7 & <Python 3.6 support? I would like to open a PR with typing hints again. It took me quite some hours to reverse-engineer `fencing.py.py` to a degree that I was able to code the fence to my satisfaction.

Comment (oalbrigt) [reply lib/fencing.py.py:38]: I think you should be able to do most of this without the changes to the library.

fence_ilo_ssh and fence_cyberpower_ssh should give some good pointers.

Comment (SchoolGuy) [reply lib/fencing.py.py:38]: Yeah, exactly my point. They give me pointers to reverse engineer but I would not expect to reverse engineer a project to contribute to it. Anyway, I will drop the typing in a minute and push again.

**Comment by oalbrigt** [comment]
Looks good. Can you run `make xml-upload` and add it to the PR, so CI doesnt fail?

File: fence-agents.spec.in [hunk 3]
```diff
@@ -55,6 +55,7 @@ fence-agents-docker \\
 fence-agents-drac \\
 fence-agents-drac5 \\
 fence-agents-eaton-snmp \\
+fence-agents-eaton-ssh \\
```
Comment (oalbrigt) [inline fence-agents.spec.in:58]: You also need to add an eaton-ssh section. Just copy the eaton-snmp section and update the description and requirements.

Comment (SchoolGuy) [reply fence-agents.spec.in:58]: Thanks for the hint. That was an oversight of me. Should be fixed.

**Comment by oalbrigt** [comment]
Thanks.

**Pull request merged**